components/tidb_query/src/codec/row/v2/encoder.rs

```diff
@@ -0,0 +1,295 @@
+            i16::MIN_I64..=i16::MAX_I64 => self.write_i16_le(v as i16),
+            i32::MIN_I64..=i32::MAX_I64 => self.write_i32_le(v as i32),
+            _ => self.write_i64_le(v),
+        }
+    }
+
+    #[allow(clippy::match_overlapping_arm)]
+    #[inline]
+    fn encode_u64(&mut self, v: u64) -> codec::Result<()> {
+        match v {
+            u8::MIN_U64..=u8::MAX_U64 => self.write_u8(v as u8),
+            u16::MIN_U64..=u16::MAX_U64 => self.write_u16_le(v as u16),
+            u32::MIN_U64..=u32::MAX_U64 => self.write_u32_le(v as u32),
+            _ => self.write_u64_le(v),
+        }
+    }
+}
+impl DatumEncoder for Vec<u8> {}
+
+struct Row {
+    is_big: bool,
+    non_null_cols: Vec<(Datum, i64)>,
+    null_cols: Vec<(Datum, i64)>,
+    value_approx_size: usize,
+}
+
+impl Row {
+    fn new(datums: Vec<Datum>, ids: &[i64]) -> Row {
+        let mut is_big = false;
+
+        let mut non_null_cols = Vec::with_capacity(ids.len());
+        let mut null_cols = Vec::with_capacity(ids.len());
+
+        let mut value_approx_size = 0;
+        for (datum, id) in datums.into_iter().zip(ids) {
+            let datum = flatten(datum);
+            if *id > 255 {
+                is_big = true;
+            }
+            value_approx_size += datum_size(&datum);
+
+            if datum == Datum::Null {
+                null_cols.push((datum, *id));
+            } else {
+                non_null_cols.push((datum, *id));
+            }
+        }
+        non_null_cols.sort_by_key(|(_, id)| *id);
+        null_cols.sort_by_key(|(_, id)| *id);
+
+        Row {
+            is_big,
+            non_null_cols,
+            null_cols,
+            value_approx_size,
+        }
+    }
+
+    // version(1) + flag(1) + columns_length(4) + ids_length + values_offset_length + values_length
+    #[inline]
+    fn approx_size(&self) -> usize {
+        6 + (self.non_null_cols.len() + self.null_cols.len()) * self.id_size()
+            + self.non_null_cols.len() * self.offset_size()
+            + self.value_approx_size
+    }
+
+    #[inline]
+    fn id_size(&self) -> usize {
+        if self.is_big {
+            LARGE_ID_SIZE
+        } else {
+            ID_SIZE
+        }
+    }
+
+    #[inline]
+    fn offset_size(&self) -> usize {
+        if self.is_big {
+            LARGE_OFFSET_SIZE
+        } else {
+            OFFSET_SIZE
+        }
+    }
+
+    #[inline]
+    fn offsets_len(&self) -> usize {
+        self.offset_size() * self.non_null_cols.len()
+    }
+}
+
+#[inline]
+fn datum_size(datum: &Datum) -> usize {
+    match *datum {
+        Datum::I64(i) => i64_size(i),
+        Datum::U64(u) => u64_size(u),
+        Datum::Bytes(ref bs) => bs.len(),
+        Datum::Dec(ref d) => d.approximate_encoded_size(),
+        Datum::Json(ref j) => j.binary_len(),
+        Datum::F64(_) | Datum::Time(_) | Datum::Dur(_) => {
+            unreachable!("these types should be flattened as numbers before")
+        }
+        _ => 0,
+    }
+}
```

Comment:
This function can be removed as long as we don't care about the performance in test utils.

Reply:
Done